src/schemapack/_internals/isolate.py

```diff
@@ -116,30 +111,25 @@ def identify_resource_dependencies(  # noqa: C901
                 context="relation resolution in datapack"
             ) from error
 
-        for target_id in target_ids:
-            if (
-                target_class_name in resource_blacklist
-                and target_id in resource_blacklist[target_class_name]
-            ):
-                continue
+        # Remove already known dependencies to avoid circular loops
+        target_ids -= _dependencies_by_class.get(target_class_name, set())
 
-            dependencies_by_class[target_class_name].add(target_id)
+        for target_id in target_ids:
+            _dependencies_by_class[target_class_name].add(target_id)
 
             # Recursively add dependencies of this target resource:
-            nested_dependencies = identify_resource_dependencies(
+            identify_resource_dependencies(
                 datapack=datapack,
                 class_name=target_class_name,
                 resource_id=target_id,
                 schemapack=schemapack,
-                _resource_blacklist=resource_blacklist,
+                _dependencies_by_class=_dependencies_by_class,
             )
-            for nested_class_name, nested_ids in nested_dependencies.items():
-                dependencies_by_class[nested_class_name].update(nested_ids)
 
-    if include_target:
-        dependencies_by_class[class_name].add(resource_id)
+    if not include_target:
+        _dependencies_by_class[class_name].discard(resource_id)
```

Comment:
Trying to trace the behavior here, but this can ever only be triggered on the first call and not when recursing downwards, if I read the code correctly:
While `include_target` is not passed along, `_dependencies_by_class` will never be `None` and thus it's still set to `True` in all subsequent calls.

All the recursion happens before and this removes the entry for the `class_name/resource_id` only for the initial call.
Could use a comment about the what and why here to make this clear.